IPython/terminal/shortcuts.py

```diff
@@ -178,6 +198,38 @@ def newline_with_copy_margin(event):
         pos_diff = cursor_start_pos - cursor_end_pos
         b.cursor_right(count=pos_diff)
 
+def newline_autoindent_outer(inputsplitter) -> Callable[..., None]:
+    """
+    Return a function suitable for inserting a indented newline after the cursor.
+
+    Fancier version of deprecated ``newline_with_copy_margin`` which should
+    compute the correct indentation of the inserted line. That is to say, indent
+    by 4 extra space after a function definition, class definition, context
+    manager... And dedent by 4 space after ``pass``, ``return``, ``raise ...``.
+    """
+
+    def newline_autoindent(event):
+        """insert a newline after the cursor indented appropriately."""
+        b = event.current_buffer
+        d = b.document
+        cursor_start_pos = b.document.cursor_position_col
+
+        if b.complete_state:
+            b.cancel_completion()
+
+        text = d.text[:d.cursor_position] + '\n'
+
+        _, indent = inputsplitter.check_complete(text)
+        b.insert_text('\n' + (' ' * (indent or 0)))
+        b.cursor_up(count=1)
+        cursor_end_pos = b.document.cursor_position_col
+        if cursor_start_pos != cursor_end_pos:
+            pos_diff = cursor_start_pos - cursor_end_pos
+            b.cursor_right(count=pos_diff)
```

Comment:
That link was: https://github.com/jonathanslenders/python-prompt-toolkit/blob/1.0.4/prompt_toolkit/buffer.py#L913